services/stats/stats.go

```diff
@@ -93,26 +97,59 @@ type RudderStatsT struct {
 	dontProcess bool
 }
 
+func getNewStatsdClientWithExpoBackoff(opts ...statsd.Option) {
+	maxWait := time.Minute * 10
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = time.Minute
+	bo.MaxElapsedTime = maxWait
+	var err error
+	if err = backoff.Retry(func() error {
+		//statsd.New returns client even when err!=nil, although that client won't work.
+		//But, since we don't want getting new client to be a blocking call. So, setting it even with the non-working client.
+		//And, will be updated once we have `err==nil`, until then stats sent to telegraf will be dropped.
+		client, err = statsd.New(opts...)
+		if err != nil {
+			isMockClient = true
+			pkgLogger.Errorf("error while setting statsd client: %v", err)
+		}
+		return err
+	}, bo); err != nil {
+		if bo.NextBackOff() == backoff.Stop {
+			//if setup didn't succeed in bo.MaxElapsedTime, then we panic. Since, if `RSERVER_ENABLE_STATS` is `true`
+			// & we failed to complete statsd setup in maxElapsedTime then something is wrong. And, we don't want to loose metrics by ignoring the error.
+			panic(err)
```

Comment:
I'm not very fond of panic-ing. I would suggest we switch to a context based cancel process, so that the server can shut down gracefully. Of course, that might require further changes in the Setup() process beyond the scope of this PR, but it something we need to keep in mind for the future and how we're handling services startup.

Reply:
make sense.

Reply:
I see that other services in main.go are already using contexts, so if it's not too much trouble let's try to add this and avoid the panic(). Otherwise, let's open a ticket to handle that separately asap.

Reply:
I've opened a ticket for this: https://www.notion.so/rudderstacks/Use-context-cancellation-instead-of-panic-to-handle-failures-during-server-startup-7fc2948a6fdb40beb46f9e0f059b098b

Reply:
IMO there is no point in panicking or shutting down gracefully. If we fail to establish a statsd connection after 15 minutes (which is the maximum elapsed time of the backoff library), then it should be about time to accept it and stop trying! Just make sure that we can handle this scenario as well, .e.g. return some error from `getNewStatsdClientWithExpoBackoff` so that the calling routine can decide how to handle the result.

Reply:
> If we fail to establish a statsd connection after 15 minutes
I believe we should keep retrying to connect forever every 15 minutes

Reply:
As an alternative, setting the `MaxElapsedTime` of the backoff to `0` would mean that we will never get a stop signal from the backoff!

Reply:
Indeed, not sending metrics is not critical to the function of the service, but it is something we should be aware of, which right now we get from the error logs printed every minute.

Reply:
there are actually many Setup() functions in main.go that requires us to pass context. Also, we are using `panic` to frequently, and most of them could be handled more gracefully, I suppose. So, I believe creating a separate ticket for this would be better.